### PR TITLE
Add all Custom Crafting classes and Tables

### PR DESCRIPTION
**Pull request by MartyParty21**
Here it is, as promised. There is a lot of thing that could have been done better, for instance Ive made a lot of possibly unnecessary classes (all the Slot//Workbenches and Containers, Guis..), but I didnt want to loose any more time on thinking how to do it. So I lost a lot of time on rewriting the classes...
However, this works. The Crafting Tables are differentiated by meta and have same ID. Each has its own container and CraftingManager tho. Only thing needed to do are all the models, textures and Guis. I set everything to look like Elven Crafting Table for now, so it looks somehow.

You can check the 6998dc8 commit to see my first version. Thats with different ID's, and its just a start. It would work in the very same way. But the same ID, latest commit is sufficient and better in my opinion.

Have a look and tell me what you think. Feel free to add some code-saving edits to solve this multi-class madness (just a question of line saving and better feeling. Everything works well as it is, so we can leave it), or anything you like!

### CODE REVIEW

File: mcp/src/minecraft/assets/minecraft/models/item/custom_crafting_table.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+    "parent": "block/custom_crafting_table"
```
Comment (scribblemaniac) [inline mcp/src/minecraft/assets/minecraft/models/item/custom_crafting_table.json]: Is this file even used? If yes, then `block/custom_crafting_table` doesn't seem to exist. If no, then we can just delete it.

File: mcp/src/minecraft/com/elementfx/tvp/ad/block/BlockCustomWorkbench.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+        public InterfaceCustomCraftingTable(World worldIn, BlockPos pos)
+        {
+        	super(worldIn, pos);
+            this.world = worldIn;
+            this.position = pos;
+        }
+
+        public ITextComponent getDisplayName()
+        {
+            return new TextComponentTranslation("tile.customWorkbench.elven.name", new Object[0]);
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/block/BlockCustomWorkbench.java]: I'm not sure where if anywhere this is used, but it doesn't seem right to leave it hard-coded to the elven workbench. Can we do something like `"tile.customWorkbench." + BlockCustomWorkbench.getUnlocalizedName(world.getBlockState(position)) + ".name"` with the getUnlocalizedName function as:
```
public static String getUnlocalizedName(IBlockState state)
{
    return ((BlockCustomWorkbench.EnumType)state.getValue(TYPE)).getUnlocalizedName();
}
```

File: mcp/src/minecraft/com/elementfx/tvp/ad/client/gui/inventory/GuiRohirrimCrafting.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import net.minecraft.client.resources.I18n;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.inventory.ContainerWorkbench;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+
+public class GuiRohirrimCrafting extends GuiContainer
+{
+	private static final ResourceLocation ROHIRRIM_CRAFTING_TABLE_GUI_TEXTURES = new ResourceLocation("textures/gui/container/crafting_table.png");
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/client/gui/inventory/GuiRohirrimCrafting.java]: Do you think that there will be any differences between the GuiXCrafting aside from the background texture and font color? If we're not going to customize them any more than that, they we can probably get away with a single class that passes those two things into the constructor.

File: mcp/src/minecraft/com/elementfx/tvp/ad/inventory/ContainerGoblinWorkbench.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+        this.onCraftMatrixChanged(this.craftMatrix);
+    }
+
+    /**
+     * Callback for when the crafting matrix is changed.
+     */
+    public void onCraftMatrixChanged(IInventory inventoryIn)
+    {
+        this.craftResult.setInventorySlotContents(0, GoblinCraftingManager.getInstance().findMatchingRecipe(this.craftMatrix, this.worldObj));
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/inventory/ContainerGoblinWorkbench.java]: Is this and the result slot type the only differences between all of the ContainerXWorkbench classes? Do you expect we'll need custom classes for other reasons?

File: mcp/src/minecraft/com/elementfx/tvp/ad/inventory/SlotGoblinCrafting.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	{
+		super(player, craftingInventory, inventoryIn, slotIndex, xPosition, yPosition);
+		this.thePlayer = player;
+        this.craftMatrix = craftingInventory;
+	}
+	
+	public void onPickupFromSlot(EntityPlayer playerIn, ItemStack stack)
+    {
+        this.onCrafting(stack);
+        ItemStack[] aitemstack = GoblinCraftingManager.getInstance().getRemainingItems(this.craftMatrix, playerIn.worldObj);
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/inventory/SlotGoblinCrafting.java]: Is this the only reason we need custom slots? Can you think of any other reason we would need custom slots in the future?

File: mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/GoblinCraftingManager.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+            public int compare(IRecipe p_compare_1_, IRecipe p_compare_2_)
+            {
+                return p_compare_1_ instanceof ShapelessRecipes && p_compare_2_ instanceof ShapedRecipes ? 1 : (p_compare_2_ instanceof ShapelessRecipes && p_compare_1_ instanceof ShapedRecipes ? -1 : (p_compare_2_.getRecipeSize() < p_compare_1_.getRecipeSize() ? -1 : (p_compare_2_.getRecipeSize() > p_compare_1_.getRecipeSize() ? 1 : 0)));
+            }
+        });
+    }
+    /**
+     * Adds a shaped recipe to the games recipe list.
+     */
+    public ShapedRecipes addRecipe(ItemStack stack, Object... recipeComponents)
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/GoblinCraftingManager.java]: Can we have a parent `CustomCraftingManager` class which contains all of these reused functions in XCraftingManager?

File: mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java [hunk 7]
```diff
@@ -10,5 +11,6 @@
 	public void addRecipes(CraftingManager manager) 
 	{
 		manager.addRecipe(new ItemStack(Items.PIPE, 1), new Object[] {"  S", "SS ", "S  ", 'S', Items.STICK});
+		manager.addShapelessRecipe(new ItemStack(Items.THROWING_STONE, 4), new Object[] {Blocks.COBBLESTONE});
```
Comment (scribblemaniac) [inline mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java:14]: This recipe doesn't seem to be working for me, but the pipe does. Can you replicate this? There may be something wrong with `addShapelessRecipe`

File: mcp/src/minecraft/net/minecraft/block/Block.java [hunk 8]
```diff
@@ -1636,14 +1636,16 @@ public static void registerBlocks()
...
         registerBlock(785, "bell", (new BlockBell()).setUnlocalizedName("bell"));
-        registerBlock(786, "elven_crafting_table", (new BlockElvenWorkbench()).setHardness(2.5F).setSoundType(SoundType.WOOD).setUnlocalizedName("elvenCraftingTable"));
-        //787 - human_crafting_table
-        //788 - gondorial_crafting_table
-        //789 - rohhirim_crafting_table
-        //790 - hobbit_crafting_table
-        //791 - mordor_crafting_table
-        //792 - isengard_crafting_table
-        //793 - goblin_crafting_table
+        Block block500 = (new BlockCustomWorkbench()).setSoundType(SoundType.WOOD).setHardness(0.8F).setUnlocalizedName("customWorkbench");
```
Comment (scribblemaniac) [inline mcp/src/minecraft/net/minecraft/block/Block.java]: Any particular reason why you chose to use two lines here? Or why the 786th block is named block500 😝

**Review by scribblemaniac — CHANGES_REQUESTED** [review]
Great stuff! It all seems to work well so far. There are a few sections that I think can be refactored to work and look a bit better and there are a few other things I want to talk about (don't worry nothing bad 😉 ) so when you get a chance and see me online, send me a message on Discord and we can discuss things further.

File: mcp/src/minecraft/assets/minecraft/models/item/custom_crafting_table.json (hunk 1 above)
Comment (MartyParty21) [reply mcp/src/minecraft/assets/minecraft/models/item/custom_crafting_table.json]: It is not used, I just used it for testing. Thanks for discovery!

File: mcp/src/minecraft/com/elementfx/tvp/ad/block/BlockCustomWorkbench.java (hunk 2 above)
Comment (MartyParty21) [reply mcp/src/minecraft/com/elementfx/tvp/ad/block/BlockCustomWorkbench.java]: You are correct. I forgot to change this also. The method itself is called when the block is right-clicked it seems. Your suggestion works just fine to solve this. 
By the way the original getUnlocalizedName method from Block.java is never overwritten. Does it mean that whenever that method is called, it returns "customWorkbench" string? And is it even needed? Only use of unlocalizedName I know about is to set the name in .lang files. And that works, if I understand corectly, thanks to this getDisplayName method.

File: mcp/src/minecraft/com/elementfx/tvp/ad/client/gui/inventory/GuiRohirrimCrafting.java (hunk 3 above)
Comment (MartyParty21) [reply mcp/src/minecraft/com/elementfx/tvp/ad/client/gui/inventory/GuiRohirrimCrafting.java]: No other diferences for sure. It should be alright with the constructors, I just wasnt sure with all the classes, and didnt want to connect them really together with constructors etc before you approve the system) I'll do it. I'll try my best with other classes as well, I wasnt too sure how to handle Crafting managers for instance, but your idea further below seems nice!

File: mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java (hunk 7 above)
Comment (MartyParty21) [reply mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java:14]: The recipe works for me well, did u use cobblestone?

File: mcp/src/minecraft/net/minecraft/block/Block.java (hunk 8 above)
Comment (MartyParty21) [reply mcp/src/minecraft/net/minecraft/block/Block.java]: All vanilla blocks which have subblocks used this :smile: but it doesnt seem to make any difference...

File: mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java (hunk 7 above)
Comment (scribblemaniac) [reply mcp/src/minecraft/com/elementfx/tvp/ad/item/crafting/RecipesCustom.java:14]: Oops, it seems I accidentally grabbed the Cobblestone Monster Egg block instead 🤦‍♂️. Works now.

**Comment by scribblemaniac** [comment]
Merged! Thanks so much for your work and patience 😁

**Pull request merged**